manual-provision-payg/testsuite/code_gen.py

```diff
@@ -0,0 +1,292 @@
+    devices = []
+    for key, value in devices_object.items():
+        block_device_obj = None
+        try:
+            block_device_obj = value["org.freedesktop.UDisks2.Block"]
+        except KeyError:
+            block_device_obj = None
+        if block_device_obj is not None:
+            if (
+                block_device_obj.get("IdUsage") == "filesystem"
+                and not block_device_obj.get("HintSystem")
+                and not block_device_obj.get("ReadOnly")
+            ):
+                devices.append(block_device_obj)
+
+    return devices
+
+
+def get_usb_path(drive):
+    user = "root"
+    root_folder = "/run/media/"
+    usb_file_path = None
+    errors = ""
+    try:
+        label = drive.get("IdLabel")
+        uuid = drive.get("IdUUID")
+
+        if label and len(label) > 0:
+            path = os.path.join(root_folder, user, label)
+            if os.path.isfile(path + "/Endless_Factory_Test.tar"):
+                usb_file_path = path
+            else:
+                errors += "couldnt find " + path + "\n"
+        else:
+            path = os.path.join(root_folder, user, uuid)
+            if os.path.isfile(path + "/Endless_Factory_Test.tar"):
+                usb_file_path = path
+            else:
+                errors += "couldnt find " + path + "\n"
+    except:
+        # TODO: Add better exception handling.
+        errors += "Error detecting USB details. \n"
+    return {"path": usb_file_path, "errors": errors}
+
+
+def password_generator(size=768, chars=string.ascii_letters + string.digits):
+    # size is currently arbitrarily set to mimic output length from apg
+    return textwrap.fill("".join(random.choice(chars) for i in range(size)))
+
+
+def write_to_file(file_location, text):
+    os.makedirs(os.path.dirname(file_location), exist_ok=True)
+    with open(file_location, "w") as file:
+        file.write(text)
+
+
+def generate_codes(key, device_id):
+    key_token_data = key_gen(key, device_id)
+    write_to_csv(key_token_data)
+    write_to_json(key_token_data)
+
+
+def generate_key():
+    key = password_generator()
+    write_to_file(cwd + "generated-key", key)
+    return key
+
+
+def backup_file(file_to_backup, backup_path):
+    # "/usr/local/share/eos-payg/key"
+    file_available = os.path.isfile(file_to_backup)
+    if file_available:
+        with open(file_to_backup, "r") as f:
+            key_contents = f.read()
+            write_to_file(backup_path, key_contents)
+
+
+def add_key():
+    key_path_destination = "/usr/local/share/eos-payg/key"
+    key_path_source = cwd + "generated-key"
+    with open(key_path_source, "r") as f:
+        generated_key_contents = f.read()
+        write_to_file(key_path_destination, generated_key_contents)
+
+
+def install_instructions():
+    config_instructions = {}
+    try:
+        with open(cwd + "config" + ".json") as f:
+            config_instructions = json.load(f)
+    except IOError:
```

Comment:
You really want `FileNotFoundError` so you don't catch another type of `IOError` inadvertently